libethereum/Client.cpp

```diff
@@ -868,19 +868,25 @@ void Client::rewind(unsigned _n)
 
 pair<h256, Address> Client::submitTransaction(TransactionSkeleton const& _t, Secret const& _secret)
 {
-    prepareForTransaction();
     TransactionSkeleton ts = populateTransactionWithDefaults(_t);
     ts.from = toAddress(_secret);
     Transaction t(ts, _secret);
+    importTransaction(t);
+    return make_pair<h256, Address>(t.sha3(), toAddress(ts.from, ts.nonce));
+}
+
+void Client::importTransaction(Transaction const& _t)
```

Comment:
`submitTransaction` and `importTransaction` are very similar in what they do, I think it would be nice if their signature is similar, so I propose to return `pair<h256, Address` here, too. Then you won't need to calculate transaction hash in the RPC layer in `eth_sendRawTransaction`.

Also, I think `Address` part of result might not be used anywhere, if so we can return just hash.